### PR TITLE
Added tab component and its stories

### PR DESCRIPTION
**Pull request by girishdudhe**
Tabs divide content into separate views which users navigate between.

- Vertical Tab

Screenshot (24)

- Horizontal Tab

Screenshot (25)

- Static Tab

Screenshot (26)

- Overflow Tab

Screenshot (28)

### CODE REVIEW

File: src/tabs/Tab.tsx [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+            </ul>
+        );
+    };
+
+    //Render Overflow Tab
+    private renderOverflowTab = (overflowTabs: TabPane[]) => {
+        const {overflowTab} = this.state;
+        let className: string | undefined;
+        if (overflowTab) {
+            overflowTab.isTabSelected ? (className = ClassNames.TABACTIVE) : (className = ClassNames.TABINACTIVE);
```
Comment (akate1) [inline src/tabs/Tab.tsx]: Please change conditional statement as above

File: src/tabs/Tab.tsx [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        );
+    };
+
+    //Render Tab bar
+    private renderTabList = (tabs: TabPane[]) => {
+        const {tabType, overflowPivot} = this.props;
+        return (
+            <ul className={ClassNames.TABMAIN}>
+                {tabs.map((tab: TabPane, index: number) => {
+                    if (tabType === TabType.OVERFLOW && overflowPivot && index < overflowPivot)
```
Comment (akate1) [inline src/tabs/Tab.tsx]: Reduce number of ifs.
what if index  > overflowPivot ?? 
   

}

File: src/tabs/Tab.tsx [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        this.setState({
+            tabs: [...tabs],
+            tabComponent: clickedTab.tabComponent,
+        });
+    };
+
+    //Render Tab Button
+    private renderTab = (tab: TabPane) => {
+        let className;
+        tab.isTabSelected ? (className = ClassNames.TABACTIVE) : (className = ClassNames.TABINACTIVE);
```
Comment (akate1) [inline src/tabs/Tab.tsx]: const className = tab.isTabSelected ? ClassNames.TABACTIVE : ClassNames.TABINACTIVE;

File: src/tabs/Tab.tsx [hunk 4]
```diff
@@ -0,0 +1,202 @@
+/**
+ * Copyright (c) 2018 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (maheshkulkarni4) [inline src/tabs/Tab.tsx]: 2019/2020 probably ?

File: src/tabs/Tab.tsx [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+ * @param {HORIZONTAL} Horizontal tabs
+ */
+export enum TabOrientation {
+    VERTICAL = "vertical",
+    HORIZONTAL = "horizontal",
+}
+
+/**
+ * Tab Types
+ * @param {STATIC} User cannot switch between tabs in this static
```
Comment (maheshkulkarni4) [inline src/tabs/Tab.tsx]: Description is bit misleading, please check, if we are can not switch between tabs, then whats the use of other tabs :)

File: src/tabs/Tab.tsx [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+ * @param {tabOrientation} orientation of tabs
+ * @param {tabType} type of tabs
+ * @param {selectedTabIndex} which tab to is selected to show first
+ * @param {overflowPivot} after how many tab remaining to put in overflow menu
+ */
+type TabProp = {
+    tabs: TabPane[];
+    tabOrientation: TabOrientation;
+    tabType: TabType;
+    selectedTabIndex?: number;
```
Comment (maheshkulkarni4) [inline src/tabs/Tab.tsx]: Variable name is not very good, does not really convey the use.

File: src/tabs/Tab.tsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    overflowPivot?: number;
```
Comment (maheshkulkarni4) [inline src/tabs/Tab.tsx]: Same for this, the props which will be used by developers should have meaningful names.

File: src/tabs/Tab.tsx [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    //Initialize tabs
+    initializeTabs = (tabs: TabPane[]) => {
+        // set index in array and selected tab
+        const {tabType, selectedTabIndex} = this.props;
+        tabs.map((tab: TabPane, index: number) => {
+            tab.index = index;
+            index === (selectedTabIndex && selectedTabIndex <= tabs.length ? selectedTabIndex : 0)
+                ? (tab.isTabSelected = true)
```
Comment (maheshkulkarni4) [inline src/tabs/Tab.tsx]: Use proper brackets to simplify the logic. Whats the use of round brackets here ?

File: src/tabs/Tab.tsx [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+                : (tab.isTabSelected = false);
+            tabType == TabType.SIMPLE
```
Comment (maheshkulkarni4) [inline src/tabs/Tab.tsx]: Please simplify the logic, its not readable enough

File: src/tabs/Tabs.stories.tsx [hunk 10]
```diff
@@ -0,0 +1,72 @@
+/**
+ * Copyright (c) 2018 Dell Inc., or its subsidiaries. All Rights Reserved.
```
Comment (maheshkulkarni4) [inline src/tabs/Tabs.stories.tsx]: Update copy right

File: src/tabs/Tabs.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+ */
+
+import * as React from "react";
+import {storiesOf} from "@storybook/react";
+import {Tab, TabOrientation, TabType} from ".";
+
+const tabsData = [
+    {
+        tabName: "Dashboard",
+        disableTab: false,
```
Comment (maheshkulkarni4) [inline src/tabs/Tabs.stories.tsx]: do we really need tab word in every prop ?  name, disabled, component -> converys same meaning.  @akate1  - please advise on this.

File: src/tabs/ClassNames.tsx [hunk 12]
```diff
@@ -0,0 +1,8 @@
+export class ClassNames {
+    public static TABMAIN: string = "nav";
+    public static TABITEM: string = "nav-item";
+    public static VERTICALTAB: string = "tabs-vertical";
+    public static TABACTIVE: string = "btn btn-link nav-link active";
+    public static TABINACTIVE: string = "btn btn-link nav-link";
+    static DROPDOWN_BTN: any;
```
[the remark this answers is not in the document]
Comment (girishdudhe) [reply src/tabs/ClassNames.tsx]: yes its unused, now removed from file.

File: src/tabs/Tab.tsx [hunk 13]
```diff
@@ -0,0 +1,202 @@
...
+ */
+export class Tab extends React.PureComponent<TabProp, TabState> {
+    constructor(props: TabProp) {
+        super(props);
+        this.initializeTabs(this.props.tabs);
+    }
+
+    //Initialize tabs
+    initializeTabs = (tabs: TabPane[]) => {
+        // set index in array and selected tab
```
[the remark this answers is not in the document]
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: Done.

File: src/tabs/Tab.tsx (hunk 3 above)
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: Done.

File: src/tabs/Tab.tsx (hunk 4 above)
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: changed to 2020.

File: src/tabs/Tabs.stories.tsx (hunk 10 above)
Comment (girishdudhe) [reply src/tabs/Tabs.stories.tsx]: Done.

File: src/tabs/Tab.tsx (hunk 1 above)
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: Done.

File: src/tabs/Tab.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: changes to selectedTabName.

File: src/tabs/Tab.tsx [hunk 14]
```diff
@@ -0,0 +1,202 @@
...
+        const {tabType, selectedTabIndex} = this.props;
+        tabs.map((tab: TabPane, index: number) => {
+            tab.index = index;
+            index === (selectedTabIndex && selectedTabIndex <= tabs.length ? selectedTabIndex : 0)
```
[the remark this answers is not in the document]
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: Done.

File: src/tabs/Tab.tsx [hunk 15]
```diff
@@ -0,0 +1,202 @@
...
+                : (tab.isTabSelected = false);
+            tabType == TabType.SIMPLE
+                ? (tab.disableTab = false)
+                : tab.disableTab
+                ? (tab.disableTab = true)
+                : (tab.disableTab = false);
+        });
+        this.state = {
+            tabs: [...tabs],
+            tabComponent: tabs[selectedTabIndex && selectedTabIndex <= tabs.length ? selectedTabIndex : 0].tabComponent,
```
[the remark this answers is not in the document]
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: I have stored is separately as it holds current visible tabs component. when tab is selected its status is changed to selected and component of that tab is set to this variable which is rendered on state change  -Aditya

File: src/tabs/Tab.tsx (hunk 2 above)
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: Done.

File: src/tabs/Tab.tsx (hunk 5 above)
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: Changed the description, it now states -> 'tabs are allowed to be disable enabled conditionally in this type.'

File: src/tabs/Tab.tsx (hunk 6 above)
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: Changed to selectedTabName which states which tab will be shown as selected initially. an optional prop, if given then tab is shown selected for that name else default first tab is shown selected.

File: src/tabs/Tab.tsx (hunk 8 above)
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: Done.

File: src/tabs/Tabs.stories.tsx (hunk 11 above)
Comment (girishdudhe) [reply src/tabs/Tabs.stories.tsx]: Done.

File: src/tabs/Tab.tsx (hunk 7 above)
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: Done.

File: src/tabs/Tab.tsx (hunk 9 above)
Comment (girishdudhe) [reply src/tabs/Tab.tsx]: Done.

File: src/tabs/Tab.tsx [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+/**
+ * props for tabs
+ * @param {index} index of tab in array
+ * @param {name} name or title of tab
+ * @param {component} React element loaded on tab selection
+ * @param {isSelected} true if tab is selected
+ * @param {isDisabled} true if tab is disabled
+ */
+type TabPane = {
+    name: string;
```
Comment (maheshkulkarni4) [inline src/tabs/Tab.tsx]: Should support custom classes as well. Add props to accept custom classes.

**Pull request merged**